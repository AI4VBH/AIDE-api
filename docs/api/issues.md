<!--
  ~ Copyright 2022 Crown Copyright
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~ http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
-->

# Issues

The `issues` endpoint provides the following APIs to retrieve issues for failed workflows.

## `GET /issues/failed?acknowledged=:acknowledged`

Returns a list of issues.

### Parameters

| Parameter Type | Label | Description | Format | Example |
|----------------|-------|-------------|--------|---------|
| Query | `:acknowledged` | `date` UTC Date ISO 8601 Date and time format. | YYYY-MM-DD | 2021-12-20

### Responses

| Code | Description |
|------|-------------|
| `200` | Issues retrieved successfully. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/issues/failed?acknowledged=2021-12-20'
```

### Example Response

```json
[
    {
        "execution_id": "981eca65-294a-434a-930a-b1828b54253a",
        "execution_time": "2022-09-23T12:50:00.305Z",
        "patient_id": "patient-id",
        "patient_name": "Lillie Dae",
        "status": "Failed",
        "task_id": "email-task",
        "workflow_instance_id": "c5e980ff-e278-441b-af64-8630bf57b7e3",
        "workflow_name": "lillie",
    }
]
```