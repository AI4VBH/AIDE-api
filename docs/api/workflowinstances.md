# Workflow Instances

## `PUT /workflwoinstances/:workflow_instance_id/executions/:execution_id/acknowledge`

Acknowledge task errors for a given execution of a workflow.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:workflow_instance_id` | The workflow instance ID |
| URL | `:execution_id` | The task execution ID |

### Response

| Code | Description |
|------|-------------|
| `200` | Returns acknowledged `MonaiWorkflowInstance` |
| `404` | Workflow instance not found or task execution not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request PUT 'http://localhost:5000/workflwoinstances/3ef05fc1-dd5e-402a-b382-a557de1e383c/executions/04533a0b-d43d-47d4-8aac-1efaca39b778/acknowledge'
```

### Example Response

```json
{
  "acknowledged_workflow_errors": "",
  "ae_title": "MONAI",
  "bucket_id": "monaideploy",
  "id": "a67a7af7-068b-44b8-a81b-def7b3e5403b",
  "input_metadata": {},
  "payload_id": "d801bd68-381d-4ff3-95a8-85c46380e075",
  "start_time": "2022-09-22T12:45:13.102Z",
  "status": "Failed",
  "tasks": [
    {
      "acknowledged_task_errors": "2022-10-03T12:00:00.001Z",
      "execution_id": "3b9d94b9-4285-45d4-bea9-491fa62b8f91",
      "execution_stats": {},
      "input_artifacts": {
        "input-dicom": "d801bd68-381d-4ff3-95a8-85c46380e075/dcm",
      },
      "input_parameters": {},
      "next_timeout": "2022-09-22T12:45:13.102Z",
      "output_artifacts": {},
      "output_directory": "d801bd68-381d-4ff3-95a8-85c46380e075/workflows/a67a7af7-068b-44b8-a81b-def7b3e5403b/3b9d94b9-4285-45d4-bea9-491fa62b8f91",
      "previous_task_id": "",
      "reason": "Connection time-out",
      "result": {},
      "status": "failed",
      "task_id": "mean-pixel-calc",
      "task_plugin_arguments": {},
      "task_start_time": "2022-09-22T12:45:13.102Z",
      "task_type": "argo",
      "timeout_interval": 0,
      "workflow_instance_id": "a67a7af7-068b-44b8-a81b-def7b3e5403b"
    }
  ],
  "workflow_id": "0c262733-05a1-4811-9d9a-b5cadb5a475a"
}
```