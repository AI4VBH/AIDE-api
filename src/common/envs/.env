PORT=5000

KEYCLOAK_URL='http://0.0.0.0:8443'
KEYCLOAK_MAIN_REALM='master'
KEYCLOAK_REALM='Demo-Realm'
KEYCLOAK_CLIENTID='nest-app'
KEYCLOAK_SECRET='83790b4f-48cd-4b6c-ac60-451a918be4b9'
KEYCLOAK_STATIC_ROLES='["admin", "clinician"]'

DATABASE_HOST='localhost'
DATABASE_NAME='nest_api'
DATABASE_USER='postgres'
DATABASE_PASSWORD='admin'
DATABASE_PORT=5432

ELASTIC_HOST=''
ELASTIC_PORT=''
ELASTIC_USE_SSL=''
ELASTIC_USERNAME=''
ELASTIC_PASSWORD=''
ELASTIC_INDEX=''

MONAI_API_HOST='https://localhost:8081/'
MIG_API_HOST='http://localhost:5010/'

MINIO_HOST='localhost'
MINIO_PORT=9001
MINIO_USE_SSL=false
MINIO_BUCKET='monaideploy'
MINIO_ACCESS_KEY='wOzmbchico52AMp3'
MINIO_SECRET_KEY='zyoT9SiMLiTCvNT0DS636FMiN4yi2258'
