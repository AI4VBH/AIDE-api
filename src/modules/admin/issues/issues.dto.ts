import { IMonaiPayload } from '../payloads/payload.interface';
import {
  MonaiWorkflowInstance,
  MonaiWorkflowTask,
} from '../workflowinstances/workflowinstances.interface';

export class IssueDto {
  task_id: string;
  status: string;
  workflow_name: string;
  patient_name: string;
  patient_id: string;
  execution_time: string;
  workflow_instance_id: string;
  execution_id: string;

  public static from(
    task: MonaiWorkflowTask,
    relativePayload: IMonaiPayload,
    relativeWorkflowInstance: MonaiWorkflowInstance,
  ) {
    const issue = new IssueDto();
    issue.task_id = task.task_id;
    issue.workflow_instance_id = task.workflow_instance_id;
    issue.execution_id = task.execution_id;
    issue.patient_id = relativePayload?.patient_details?.patient_id;
    issue.patient_name = relativePayload?.patient_details?.patient_name;
    issue.status = task.status;
    issue.execution_time = task.task_start_time;
    issue.workflow_name = relativeWorkflowInstance.workflow_name;
    return issue;
  }

  // TODO: Implement fromEntity method
  // public static fromEntity(entity: Issue) {
  //   return this.from({
  //     task_id: entity.id,
  //   });
  // }

  // TODO: Implement toEntity method
  // public toEntity(issue: Issue = null) {
  //   const it = new Issue();
  //   it.task_id = this.task_id;
  //   return it;
  // }
}
