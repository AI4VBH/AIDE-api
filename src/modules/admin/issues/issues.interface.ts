export interface Task {
  task_id: number;
  status: string;
  model_name: string;
  patient_name: string;
  patient_id: string;
  execution_time: string;
}
