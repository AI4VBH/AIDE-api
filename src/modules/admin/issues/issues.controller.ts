import { Controller, Get, Patch } from '@nestjs/common';
import { Issue } from './issues.interface';
import { IssuesService } from './issues.service';

@Controller('issues')
export class IssuesController {
  constructor(private readonly appService: IssuesService) {}

  @Get()
  getHello(): Issue[] {
    return this.appService.getIssues();
  }

  @Patch()
  dismissTask(): Issue {
    return this.appService.dismissTask();
  }
}
