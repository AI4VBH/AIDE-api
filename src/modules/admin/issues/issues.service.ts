import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { AxiosResponse } from 'axios';
import { lastValueFrom } from 'rxjs';
import { IMonaiPayload } from '../payloads/payload.interface';
import { MonaiWorkflowInstance } from '../workflowinstances/workflowinstances.interface';
import { IssueDto } from './issues.dto';

@Injectable()
export class IssuesService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  async getIssues(
    workflowInstances: MonaiWorkflowInstance[],
  ): Promise<IssueDto[]> {
    const payloadIds = [
      ...new Set(workflowInstances.map((wf) => wf.payload_id)),
    ];

    const promises: Promise<AxiosResponse<IMonaiPayload>>[] = [];

    for (const payloadId of payloadIds) {
      promises.push(
        lastValueFrom(
          this.httpService.get<IMonaiPayload>(`payloads/${payloadId}`),
        ),
      );
    }

    const payloads = (await Promise.all(promises)).map((resp) => resp.data);

    return this.mapIssues(payloads, workflowInstances);
  }

  private mapIssues(
    payloads: IMonaiPayload[],
    workflowInstances: MonaiWorkflowInstance[],
  ): IssueDto[] {
    const issues: IssueDto[] = [];

    for (const workflowInstance of workflowInstances) {
      const relativePayload = payloads.find(
        (p) => p.payload_id == workflowInstance.payload_id,
      );

      const wfiIssues = workflowInstance.tasks.map((task) =>
        IssueDto.from(task, relativePayload, workflowInstance),
      );

      issues.push(...wfiIssues);
    }

    return issues;
  }
}
