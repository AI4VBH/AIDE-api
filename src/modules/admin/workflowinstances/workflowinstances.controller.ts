import { Controller, Put, Inject, Param, UseFilters } from '@nestjs/common';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import { WorkflowInstancesService } from './workflowinstances.service';

@Controller('workflowinstances')
@UseFilters(ExternalServerExceptionFilter)
export class WorkflowInstanceController {
  @Inject(WorkflowInstancesService)
  private readonly wfiService: WorkflowInstancesService;

  @Put(':workflow_instance_id/executions/:execution_id/acknowledge')
  acknowledgeTaskError(
    @Param('workflow_instance_id') workflow_instance_id,
    @Param('execution_id') execution_id,
  ) {
    return this.wfiService.acknowledgeTaskError(
      workflow_instance_id,
      execution_id,
    );
  }
}
