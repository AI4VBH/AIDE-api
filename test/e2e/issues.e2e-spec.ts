import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { HttpModule } from '@nestjs/axios';
import { ConfigModule } from '@nestjs/config';
import * as request from 'supertest';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { HttpConfigService } from 'shared/http/http.service';
import IssuesMocks from '../test_data/mocks/issues/issues';
import { IssuesController } from 'modules/admin/issues/issues.controller';
import { IssuesService } from 'modules/admin/issues/issues.service';
import { WorkflowInstancesService } from 'modules/admin/workflowinstances/workflowinstances.service';
import { WorkflowInstanceController } from 'modules/admin/workflowinstances/workflowinstances.controller';

const server = setupServer();
const testMonaiBasePath = 'https://localhost:7337';

describe('/issues integration Tests', () => {
  let app: INestApplication;

  beforeAll(() => {
    server.listen({
      onUnhandledRequest: 'bypass',
    });
    server.printHandlers();
  });

  afterAll(() => {
    server.close();
  });

  beforeEach(async () => {
    server.resetHandlers();

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forRoot({
          isGlobal: true,
          // we mock the path to the MONAI API
          load: [
            () => ({
              MONAI_API_HOST: testMonaiBasePath,
            }),
          ],
        }),
        // we register the HttpService with no stubbing! ðŸŽ‰
        HttpModule.registerAsync({
          useClass: HttpConfigService,
        }),
      ],
      controllers: [IssuesController, WorkflowInstanceController],
      providers: [IssuesService, WorkflowInstancesService],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });

  it('(GET) issues/failed?acknowledged=YYYY-MM-DD - Multiple workflow instances with 1 failed task each', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/failed?acknowledged=2022-01-01`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.failedTasks1));
        },
      ),
      rest.get(
        `${testMonaiBasePath}/payloads/fea1bc88-73e3-436a-aa6e-83364ff8d7d1`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.payload1));
        },
      ),
      rest.get(
        `${testMonaiBasePath}/payloads/a0e7b480-1bd8-41f9-a6cf-1ab7952aab32`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.payload2));
        },
      ),
      rest.get(
        `${testMonaiBasePath}/payloads/3147c4b5-6652-4d25-82f8-f9ed58b74e43`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.payload3));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/issues/failed?acknowledged=2022-01-01',
    );
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  it('(GET) issues/failed?acknowledged=YYYY-MM-DD - Multiple workflow instances with 1 failed task each, with time', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/failed?acknowledged=2022-01-01T01:01:01`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.failedTasks1));
        },
      ),
      rest.get(
        `${testMonaiBasePath}/payloads/fea1bc88-73e3-436a-aa6e-83364ff8d7d1`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.payload1));
        },
      ),
      rest.get(
        `${testMonaiBasePath}/payloads/a0e7b480-1bd8-41f9-a6cf-1ab7952aab32`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.payload2));
        },
      ),
      rest.get(
        `${testMonaiBasePath}/payloads/3147c4b5-6652-4d25-82f8-f9ed58b74e43`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.payload3));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/issues/failed?acknowledged=2022-01-01T01:01:01',
    );
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  it('(GET) issues/failed?acknowledged=YYYY-MM-DD - Single workflow instance with multiple failed tasks', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/failed?acknowledged=2022-01-01`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.failedTasks2));
        },
      ),
      rest.get(
        `${testMonaiBasePath}/payloads/fea1bc88-73e3-436a-aa6e-83364ff8d7d1`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.payload1));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/issues/failed?acknowledged=2022-01-01',
    );
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  it('(GET) issues/failed?acknowledged=YYYY-MM-DD - Single workflow instance with multiple failed tasks, with time', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/failed?acknowledged=2022-01-01T01:01:01`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.failedTasks2));
        },
      ),
      rest.get(
        `${testMonaiBasePath}/payloads/fea1bc88-73e3-436a-aa6e-83364ff8d7d1`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.payload1));
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/issues/failed?acknowledged=2022-01-01T01:01:01',
    );
    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot();
  });

  it('(GET) issues/failed?acknowledged=YYYY-MM-DD - No workflow instances returned', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/failed`,
        (_request, response, context) => {
          return response(
            context.status(404),
            context.json(IssuesMocks.noUnacknowledgedTasksError),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/issues/failed?acknowledged=2022-01-01',
    );
    expect(response.status).toBe(404);
    expect(response.body).toMatchSnapshot();
  });

  it('(GET) issues/failed?acknowledged=YYYY-MM-DD - Workflow instance exists but no matching payload', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/failed?acknowledged=2022-01-01`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.failedTasks2));
        },
      ),
      rest.get(
        `${testMonaiBasePath}/payloads/fea1bc88-73e3-436a-aa6e-83364ff8d7d1`,
        (_request, response, context) => {
          return response(
            context.status(404),
            context.json(IssuesMocks.noMatchingPayload),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/issues/failed?acknowledged=2022-01-01',
    );
    expect(response.status).toBe(404);
    expect(response.body).toMatchSnapshot();
  });

  it('(GET) issues/failed?acknowledged=YYYY-MM-DD - Invalid date, workflow instance date in the future', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/failed?acknowledged=2023-01-01`,
        (_request, response, context) => {
          return response(
            context.status(400),
            context.json(IssuesMocks.invalidDate),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/issues/failed?acknowledged=2023-01-01',
    );
    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });

  it('(GET) issues/failed?acknowledged=YYYY-MM-DD - Payload id is invalid GUID', async () => {
    server.use(
      rest.get(
        `${testMonaiBasePath}/workflowinstances/failed?acknowledged=2022-01-01`,
        (_request, response, context) => {
          return response(context.json(IssuesMocks.invalidPayloadId));
        },
      ),
      rest.get(
        `${testMonaiBasePath}/payloads/invalid-guid`,
        (_request, response, context) => {
          return response(
            context.status(400),
            context.json(IssuesMocks.invalidPayloadError),
          );
        },
      ),
    );
    const response = await request(app.getHttpServer()).get(
      '/issues/failed?acknowledged=2022-01-01',
    );
    expect(response.status).toBe(400);
    expect(response.body).toMatchSnapshot();
  });

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) issues/failed?acknowledged=YYYY-MM-DD - Generic MONAI error returned on workflow instances request',
    async (code) => {
      server.use(
        rest.get(
          `${testMonaiBasePath}/workflowinstances/failed?acknowledged=2022-01-01`,
          (_request, response, context) => {
            return response(context.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer()).get(
        '/issues/failed?acknowledged=2022-01-01',
      );
      expect(response.status).toBe(500);
      expect(response.body).toMatchSnapshot();
    },
  );

  it.each([408, 500, 501, 502, 503, 504])(
    '(GET) issues/failed?acknowledged=YYYY-MM-DD - Generic MONAI error returned on payload request',
    async (code) => {
      server.use(
        rest.get(
          `${testMonaiBasePath}/workflowinstances/failed?acknowledged=2022-01-01`,
          (_request, response, context) => {
            return response(context.json(IssuesMocks.failedTasks2));
          },
        ),
        rest.get(
          `${testMonaiBasePath}/payloads/fea1bc88-73e3-436a-aa6e-83364ff8d7d1`,
          (_request, response, context) => {
            return response(context.status(code));
          },
        ),
      );
      const response = await request(app.getHttpServer()).get(
        '/issues/failed?acknowledged=2022-01-01',
      );
      expect(response.status).toBe(500);
      expect(response.body).toMatchSnapshot();
    },
  );
});
