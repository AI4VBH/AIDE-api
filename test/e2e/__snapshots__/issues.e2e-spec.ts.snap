// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/issues integration Tests (GET) issues/failed?acknowledged=YYYY-MM-DD - Invalid date, workflow instance date in the future 1`] = `
{
  "detail": "Failed to validate date, date is in the future",
  "instance": "/workflowinstances/failed?acknowledged=2023-01-01",
  "status": 400,
  "title": "Not Found",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/issues integration Tests (GET) issues/failed?acknowledged=YYYY-MM-DD - Multiple workflow instances with 1 failed task each 1`] = `
[
  {
    "execution_id": "1f3d4b42-8a8f-41b1-9cc7-bfb19d9716d1",
    "execution_time": "2022-01-01T01:01:01.001Z",
    "patient_id": "1d0253c4-8fab-41df-a414-55d52e4c6c31",
    "patient_name": "Alex Bazin payload 1",
    "status": "Failed",
    "task_id": "test-task-id-1",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e1",
  },
  {
    "execution_id": "70613bf7-d647-48e2-a59c-5edece8ddc22",
    "execution_time": "2022-01-01T01:01:01.002Z",
    "patient_id": "b3f774a4-358f-444e-a5d1-2d6087869e62",
    "patient_name": "Alex Bazin payload 2",
    "status": "Failed",
    "task_id": "test-task-id-2",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e2",
  },
  {
    "execution_id": "d849eb4b-1bd7-4cfc-a8ec-b94998422663",
    "execution_time": "2022-01-01T01:01:01.003Z",
    "patient_id": "fce96020-7105-4773-b4dd-6f2377984cb3",
    "patient_name": "Alex Bazin payload 3",
    "status": "Failed",
    "task_id": "test-task-id-3",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e3",
  },
]
`;

exports[`/issues integration Tests (GET) issues/failed?acknowledged=YYYY-MM-DD - Multiple workflow instances with 1 failed task each, with time 1`] = `
[
  {
    "execution_id": "1f3d4b42-8a8f-41b1-9cc7-bfb19d9716d1",
    "execution_time": "2022-01-01T01:01:01.001Z",
    "patient_id": "1d0253c4-8fab-41df-a414-55d52e4c6c31",
    "patient_name": "Alex Bazin payload 1",
    "status": "Failed",
    "task_id": "test-task-id-1",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e1",
  },
  {
    "execution_id": "70613bf7-d647-48e2-a59c-5edece8ddc22",
    "execution_time": "2022-01-01T01:01:01.002Z",
    "patient_id": "b3f774a4-358f-444e-a5d1-2d6087869e62",
    "patient_name": "Alex Bazin payload 2",
    "status": "Failed",
    "task_id": "test-task-id-2",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e2",
  },
  {
    "execution_id": "d849eb4b-1bd7-4cfc-a8ec-b94998422663",
    "execution_time": "2022-01-01T01:01:01.003Z",
    "patient_id": "fce96020-7105-4773-b4dd-6f2377984cb3",
    "patient_name": "Alex Bazin payload 3",
    "status": "Failed",
    "task_id": "test-task-id-3",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e3",
  },
]
`;

exports[`/issues integration Tests (GET) issues/failed?acknowledged=YYYY-MM-DD - No workflow instances returned 1`] = `
{
  "detail": "Request failed, no workflow instances found since 2022-01-01",
  "instance": "/workflowinstances/failed?acknowledged=2022-01-01",
  "status": 404,
  "title": "Not Found",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/issues integration Tests (GET) issues/failed?acknowledged=YYYY-MM-DD - Payload id is invalid GUID 1`] = `
{
  "detail": "Request failed, payload is not a valid GUID",
  "instance": "/payloads/fea1bc88-73e3-436a-aa6e-83364ff8d7d1",
  "status": 400,
  "title": "Not Found",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;

exports[`/issues integration Tests (GET) issues/failed?acknowledged=YYYY-MM-DD - Single workflow instance with multiple failed tasks 1`] = `
[
  {
    "execution_id": "1f3d4b42-8a8f-41b1-9cc7-bfb19d9716d1",
    "execution_time": "2022-01-01T01:01:01.001Z",
    "patient_id": "1d0253c4-8fab-41df-a414-55d52e4c6c31",
    "patient_name": "Alex Bazin payload 1",
    "status": "Failed",
    "task_id": "test-task-id-1",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e1",
  },
  {
    "execution_id": "70613bf7-d647-48e2-a59c-5edece8ddc22",
    "execution_time": "2022-01-01T01:01:01.002Z",
    "patient_id": "1d0253c4-8fab-41df-a414-55d52e4c6c31",
    "patient_name": "Alex Bazin payload 1",
    "status": "Failed",
    "task_id": "test-task-id-2",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e1",
  },
  {
    "execution_id": "d849eb4b-1bd7-4cfc-a8ec-b94998422663",
    "execution_time": "2022-01-01T01:01:01.003Z",
    "patient_id": "1d0253c4-8fab-41df-a414-55d52e4c6c31",
    "patient_name": "Alex Bazin payload 1",
    "status": "Failed",
    "task_id": "test-task-id-3",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e1",
  },
]
`;

exports[`/issues integration Tests (GET) issues/failed?acknowledged=YYYY-MM-DD - Single workflow instance with multiple failed tasks, with time 1`] = `
[
  {
    "execution_id": "1f3d4b42-8a8f-41b1-9cc7-bfb19d9716d1",
    "execution_time": "2022-01-01T01:01:01.001Z",
    "patient_id": "1d0253c4-8fab-41df-a414-55d52e4c6c31",
    "patient_name": "Alex Bazin payload 1",
    "status": "Failed",
    "task_id": "test-task-id-1",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e1",
  },
  {
    "execution_id": "70613bf7-d647-48e2-a59c-5edece8ddc22",
    "execution_time": "2022-01-01T01:01:01.002Z",
    "patient_id": "1d0253c4-8fab-41df-a414-55d52e4c6c31",
    "patient_name": "Alex Bazin payload 1",
    "status": "Failed",
    "task_id": "test-task-id-2",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e1",
  },
  {
    "execution_id": "d849eb4b-1bd7-4cfc-a8ec-b94998422663",
    "execution_time": "2022-01-01T01:01:01.003Z",
    "patient_id": "1d0253c4-8fab-41df-a414-55d52e4c6c31",
    "patient_name": "Alex Bazin payload 1",
    "status": "Failed",
    "task_id": "test-task-id-3",
    "workflow_instance_id": "e26bec36-9a22-4be2-8402-379cb96c22e1",
  },
]
`;

exports[`/issues integration Tests (GET) issues/failed?acknowledged=YYYY-MM-DD - Workflow instance exists but no matching payload 1`] = `
{
  "detail": "Request failed, no matching payload found",
  "instance": "/payloads/fea1bc88-73e3-436a-aa6e-83364ff8d7d1",
  "status": 404,
  "title": "Not Found",
  "traceId": "00-1a488bac0cc6ae854a5b6ab0c237dd8f-5ab3204912c13ca3-00",
  "type": "https://tools.ietf.org/html/rfc7231#section-6.5.4",
}
`;
